core/src/main/java/io/grpc/internal/TransportSet2.java

```diff
@@ -0,0 +1,501 @@
+        }
+      });
+  }
+
+  void shutdownNow(Status reason) {
+    shutdown();
+    Collection<ManagedClientTransport> transportsCopy;
+    synchronized (lock) {
+      transportsCopy = new ArrayList<ManagedClientTransport>(transports);
+    }
+    for (ManagedClientTransport transport : transportsCopy) {
+      transport.shutdownNow(reason);
+    }
+  }
+
+  @GuardedBy("lock")
+  private void cancelReconnectTask() {
+    if (reconnectTask != null) {
+      reconnectTask.cancel(false);
+      reconnectTask = null;
+    }
+  }
+
+  @Override
+  public String getLogId() {
+    return GrpcUtil.getLogId(this);
+  }
+
+  @VisibleForTesting
+  ConnectivityState getState() {
+    synchronized (lock) {
+      return state.getState();
+    }
+  }
+
+  /** Listener for real transports. */
+  private class TransportListener implements ManagedClientTransport.Listener {
+    final ManagedClientTransport transport;
+    final SocketAddress address;
+
+    TransportListener(ManagedClientTransport transport, SocketAddress address) {
+      this.transport = transport;
+      this.address = address;
+    }
+
+    @Override
+    public void transportReady() {
+      if (log.isLoggable(Level.FINE)) {
+        log.log(Level.FINE, "[{0}] {1} for {2} is ready",
+            new Object[] {getLogId(), transport.getLogId(), address});
+      }
+      ConnectivityState savedState;
+      synchronized (lock) {
+        savedState = state.getState();
+        reconnectPolicy = null;
+        nextAddressIndex = 0;
+        if (savedState == SHUTDOWN) {
+          // activeTransport should have already been set to null by shutdown(). We keep it null.
+          Preconditions.checkState(activeTransport == null,
+              "Unexpected non-null activeTransport");
+        } else if (pendingTransport == transport) {
+          gotoNonErrorState(READY);
+          activeTransport = transport;
+          pendingTransport = null;
+        }
+      }
+      if (savedState == SHUTDOWN) {
+        transport.shutdown();
+      }
+    }
+
+    @Override
+    public void transportInUse(boolean inUse) {
+      handleTransportInUseState(transport, inUse);
+    }
+
+    @Override
+    public void transportShutdown(Status s) {
+      boolean allAddressesFailed = false;
+      if (log.isLoggable(Level.FINE)) {
+        log.log(Level.FINE, "[{0}] {1} for {2} is being shutdown with status {3}",
+            new Object[] {getLogId(), transport.getLogId(), address, s});
+      }
+      Runnable runnable = null;
+      synchronized (lock) {
+        if (state.getState() == SHUTDOWN) {
+          return;
+        }
+        if (activeTransport == transport) {
+          gotoNonErrorState(IDLE);
+          activeTransport = null;
+        } else if (pendingTransport == transport) {
+          // Continue reconnect if there are still addresses to try.
+          if (nextAddressIndex == 0) {
+            allAddressesFailed = true;
```

Comment:
I'm not sure that's a good idea, but can't we call `scheduleBackoff()` here? The method anyway acquires `lock`. We could then kill the special case for `if (state.getState() == SHUTDOWN)` since the state should be `CONNECTING`.